src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -416,7 +428,54 @@ private void RenderUnboundAttribute(string attributeName, Chunk attributeValueCh
                 // C# code, then we need to buffer it.
                 if (!isPlainTextValue)
                 {
-                    BuildBufferedWritingScope(attributeValueChunk, htmlEncodeValues: true);
+                    BuildBufferedWritingScope(renderValue: () =>
+                    {
+                        IChunkVisitor visitor;
+                        var parentChunk = attributeValueChunk as ParentChunk;
+                        containsDynamicCodeAttributeChunk = parentChunk
+                            .Children
+                            .Any(child => child is DynamicCodeAttributeChunk);
+
+                        if (!_designTimeMode && containsDynamicCodeAttributeChunk)
+                        {
+                            // If there are any non-dynamic code attribute chunks at the top level of the attribute
+                            // value's chunk it means there's static content in the value, the attribute will always
+                            // render.
+                            forceRenderAttribute = parentChunk
+                                .Children
+                                .Any(child => !(child is DynamicCodeAttributeChunk));
+
+                            if (!forceRenderAttribute)
+                            {
+                                _writer
+                                    .WriteStartAssignment(ShouldRenderTagHelperAttributeVariableName)
+                                    .WriteBooleanLiteral(value: false)
+                                    .WriteEndLine();
+                            }
+
+                            _writer
+                                .WriteStartAssignment(OriginalTagHelperAttributeValueVariableName)
+                                .WriteStartNewObject(_tagHelperContext.UnchangedTagHelperAttributeValueBufferTypeName)
+                                .Write(_tagHelperContext.OutputTextWriterPropertyName)
+                                .Write(".")
+                                .Write(nameof(TextWriter.Encoding))
+                                .WriteLine(");");
+
+                            visitor = new UnboundDynamicAttributeValueVisitor(
+                                attributeName,
+                                forceRenderAttribute,
```

Comment:
Since you're passing this in, moving lines [448](https://github.com/aspnet/Razor/pull/446/files#diff-6f3af99214f3f3e7141c9d7bd2aea6c0R448) through [462](https://github.com/aspnet/Razor/pull/446/files#diff-6f3af99214f3f3e7141c9d7bd2aea6c0R462) into `UnboundDynamicAttributeValueVisitor`. This block would need to be refactored otherwise.